### DISCUSSION TITLE
New script: list of recently selected nodes, with navigation capabilities

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#1752] [Show and tell]
THE MOST RECENT VERSION IS IN THE REPOSITORY: [UtilityPanels](https://github.com/euu2021/Freeplane_UtilityPanels)
---

I always wanted this, and I remember it being requested in the forum:

![explorer_b9xQM1sszp](https://github.com/freeplane/freeplane/assets/77707706/d2ba5769-8d90-43ff-8f29-1eaf17b1edba)

Some design choices that I made, but can be changed:
- The list is limited to 10 items;
- The listener only exists while the list window exists;
- The node is not added again to the list, if it is already in the list;

Now, I want to integrate it to the FP map view, like the MapOverview. Ideally, with a transparent background. As you can see in the code, I've made some tests with transparent background, but I want it integrated in the UI, replicated in each map view. But, it's not easy...

<details>
  <summary>The script:</summary>
  

```
import javax.swing.*
import javax.swing.event.ListSelectionListener
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.awt.Color

import org.freeplane.features.mode.Controller
import org.freeplane.features.map.NodeModel
import org.freeplane.features.map.INodeSelectionListener

counter = 0
history = []

INodeSelectionListener listener = new INodeSelectionListener() {
    @Override
    public void onDeselect(final NodeModel node) {
    }

    @Override
    public void onSelect(final NodeModel node) {
        if (history.contains(node)) {return}
        history.add(node)
        if (history.size() > 10) {
            history.remove(0)}
        else{}
        //c.statusInfo = history.size()
        updateGui(history)
    }
}

Controller.currentController.modeController.mapController.addNodeSelectionListener(listener)

JFrame frame 

def initializeGui() {
    SwingUtilities.invokeLater {
        frame = new JFrame('Node List GUI')
        frame.setSize(300, 400)
        //frame.setUndecorated(true) // Set frame undecorated
        frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE)
        //frame.setBackground(new Color(0,0,0,1)) // Make frame background transparent
		
        frame.addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                Controller.currentController.modeController.mapController.nodeSelectionListeners.each {
                    try {
                        if (it.history) {
                            Controller.currentController.modeController.mapController.removeNodeSelectionListener(it)
                        }
                    } catch(Exception ex) {
                    }
                }
            }
        })
        
        updateGui(history) // Initialize GUI components
        frame.setLocationRelativeTo(null) // Center the frame
        frame.setVisible(true)
    }
}

def updateGui(List<NodeModel> nodes) {
    frame.getContentPane().removeAll()
    //frame.getContentPane().setBackground(new Color(0,0,0,0)) // Make content pane background transparent
    //frame.getContentPane().setOpaque(false) // Make content pane non-opaque
    
    def listModel = nodes.collect { it.text }

    JList<String> list = new JList(listModel.toArray())
    //list.setOpaque(false) // Make JList non-opaque
    //list.setBackground(new Color(0,0,0,0)) // Make JList background transparent
    
    JScrollPane scrollPane = new JScrollPane(list)
    scrollPane.setOpaque(false) // Make JScrollPane non-opaque
    //scrollPane.getViewport().setOpaque(false) // Make JScrollPane viewport non-opaque
    //scrollPane.setBackground(new Color(0,0,0,0)) // Make JScrollPane background transparent

    list.addListSelectionListener({ e ->
        if (!e.getValueIsAdjusting()) {
            Integer selectedItemIndex = list.getSelectedIndex()
            selectedItemNodeId = history[selectedItemIndex].id
            nodeToBeSelected = Controller.currentController.mapViewManager.mapView.getMapSelection().getMap().getNodeForID(selectedItemNodeId)
            Controller.currentController.mapViewManager.mapView.getMapSelection().selectAsTheOnlyOneSelected(nodeToBeSelected)
        }
    } as ListSelectionListener)

    frame.getContentPane().add(scrollPane)
    frame.getContentPane().validate()
    frame.repaint()
}

initializeGui()
```

</details>

**Comment by Illegit** [comment]
Hi euu2021, interesting, as you want to develop it further please even add this topic to be solved: if you change the map it doesn't follow, it stays only in the current map but cannot open in the previous map.

**Reply by euu2021** [reply]
> Hi euu2021, interesting, as you want to develop it further please even add this topic to be solved: if you change the map it doesn't follow, it stays only in the current map but cannot open in the previous map.

I see, the ideal would be to have options for multiple maps. For example, choosing to keep one single list for all maps, or different lists for each map. Perhaps tabs that would automatically change according to the active map, but would allow the user to manually change the tab. Or even a symbol before each item of the list, to show if that node is not from the active map.

**Comment by euu2021** [comment]
This script is getting neat. Now, it gets embedded in the FP window; follows the active map view; follows the colors of the actual node:
![explorer_vZMgZHTc5l](https://github.com/freeplane/freeplane/assets/77707706/3e69bea7-79de-4e55-8608-31c622f0a2f2)

There are a couple more things that I want: pinning nodes (maybe a horizontal line at the top of the screen); drag and drop operations directly with the nodes of the map. 

Currently, the easiest way to deactivate the script is restarting FP.

<details>
  <summary>The script:</summary>
  
  
  ```groovy
  import javax.swing.*
import java.awt.*
import java.util.List
import javax.swing.event.ListSelectionListener
import java.awt.Component

import org.freeplane.features.mode.Controller
import org.freeplane.features.map.NodeModel
import org.freeplane.features.map.INodeSelectionListener
import org.freeplane.features.nodestyle.NodeStyleController
import org.freeplane.features.styles.LogicalStyleController.StyleOption
import org.freeplane.features.ui.IMapViewChangeListener

List<NodeModel> history = []

JScrollPane parentPanel
JPanel childPanel

createPanels()

INodeSelectionListener listener = new INodeSelectionListener() {
    @Override
    public void onDeselect(NodeModel node) {
        SwingUtilities.invokeLater { updateGui(history) }
    }

    @Override
    public void onSelect(NodeModel node) {
        parentPanel
        childPanel
        if (!history.contains(node)) {
            history.add(node)
            if (history.size() > 10) {
                history.remove(0)
            }
            SwingUtilities.invokeLater { updateGui(history) }
        }
    }
}

createdListener = listener

Controller.currentController.modeController.mapController.addNodeSelectionListener(listener)

SwingUtilities.invokeLater { updateGui(history) }

Controller.currentController.mapViewManager.addMapViewChangeListener(new IMapViewChangeListener() {
    public void afterViewChange(final Component oldView, final Component newView) {
        if (newView == null) {
            return;
        }
        createPanels()
    }
})

def createPanels(){
    parentPanel = Controller.currentController.mapViewManager.mapView.parent.parent as JScrollPane
    Dimension parentSize = parentPanel.getSize() 

    childPanel = new JPanel(new BorderLayout())
    childPanel.setBackground(new Color(0, 0, 0, 10)) 
    int childWidth = 80
    int childHeight = 130

    childPanel.setBounds(0, 0, childWidth, childHeight)

    parentPanel.add(childPanel)
    parentPanel.setComponentZOrder(childPanel, 0);
    parentPanel.revalidate()
    parentPanel.repaint()
}

def updateGui(List<NodeModel> nodes) {
    childPanel.removeAll()

    DefaultListModel<NodeModel> listModel = new DefaultListModel<>()
    nodes.each { listModel.addElement(it) }
    JList<NodeModel> jList = new JList<>(listModel)

    JScrollPane scrollPane = new JScrollPane(jList)
    jList.setOpaque(true)
    scrollPane.setOpaque(true)
    scrollPane.getViewport().setOpaque(true)

    scrollPane.setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_NEVER)
    scrollPane.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_NEVER)

    jList.setCellRenderer(new DefaultListCellRenderer() {
        @Override
        public Component getListCellRendererComponent(JList<?> list, Object value, int index, boolean isSelected, boolean cellHasFocus) {
            JLabel label = (JLabel) super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus)
            NodeModel nodeItem = (NodeModel) value 

            Color nodeBackgroundColor = NodeStyleController.getController().getBackgroundColor(nodeItem, StyleOption.FOR_UNSELECTED_NODE)
            Color nodeFontColor = NodeStyleController.getController().getColor(nodeItem, StyleOption.FOR_UNSELECTED_NODE)

            if (isSelected) {
                label.setBackground(list.getSelectionBackground())
                label.setForeground(list.getSelectionForeground())
            } else {
                label.setBackground(nodeBackgroundColor)
                label.setForeground(nodeFontColor)
            }

            return label
        }
    })

    jList.addListSelectionListener({ e ->
        if (!e.getValueIsAdjusting()) {
            Integer selectedItemIndex = jList.getSelectedIndex()
            NodeModel selectedItemNode = nodes[selectedItemIndex]
            Controller.currentController.mapViewManager.mapView.getMapSelection().selectAsTheOnlyOneSelected(selectedItemNode)
        }
    } as ListSelectionListener)

    childPanel.add(scrollPane, BorderLayout.CENTER)
    childPanel.revalidate()
    childPanel.repaint()
}

  ```
  
</details>

**Comment by dpolivaev** [comment]
What do you think about adding the feature directly to Freeplane and sending me a pull request?

**Reply by euu2021** [reply]
Nice! I will spend more time testing it to see how it behaves in real life situations, and then I create a pull request. 

I want to try adding other features, but I pull a simple and neat implementation, with just the list of recent nodes, as it is probably of more general interest.

**Comment by Illegit** [comment]
Hi euu2021, here is a screenshot: 
![image](https://github.com/freeplane/freeplane/assets/62674157/7495ead1-65b8-447a-89d1-c4538f726a39)
As you can see the list is too small. Is it possible to add the same features of the map overview to resize it and relocate it?

**Reply by Illegit** [reply]
the list is in the upper left corner and close to it there is the map overview

**Reply by euu2021** [reply]
> As you can see the list is too small. Is it possible to add the same features of the map overview to resize it and relocate it?

Yes, it is necessary to improve those layout situations. But, I think it is too difficult for me. I will leave it to Dimitry to do the polishing.

**Reply by euu2021** [reply]
And the problem is also related to the Look and Feel (Preferences…->Appearance->Look and feel->Look and Feel). Some of them, like Darcula and Atom One make the sizes of the window change.

**Comment by euu2021** [comment]
I created a repository for this is script, and I will call it UtilityPanels. I included some new things: https://github.com/euu2021/Freeplane_UtilityPanels

@Illegit I think I solved the problem of the size of the font changing according to the Look and Feel. You can set the sizes for the panels and font, according to your needs, at those lines:

```
int childWidth = 80
int childHeight = 170

int pinnedPanelHeight = 130

Font font = new Font("Dialog", Font.PLAIN, 10)
```

**Reply by Illegit** [reply]
Thanks euu2021, much better. I reset the sizes and it is readable. May I highlight some issues?
![image](https://github.com/freeplane/freeplane/assets/62674157/9e5f6a7b-9a40-4b53-b10e-ed64c4cdba50)
1) you can see that the new icons are not present (errors in the panels)
2) you can see that the currently selected node is in the panel but not visible
In any case the script is working and could be already usable.

**Reply by euu2021** [reply]
> you can see that the new icons are not present (errors in the panels)

For some reason, your PC is not rendering the symbols. Try changing the font at line:
Font font = new Font("Dialog", Font.PLAIN, 10)

Or, picking different symbols at lines:
label.setText("📌" + label.getText())}
label.setText("⚠|" + label.getText())

> you can see that the currently selected node is in the panel but not visible

You need to increase the height a little bit more at line, so the panel shows all 10 items:
int childHeight = 170

**Reply by euu2021** [reply]
> you can see that the new icons are not present (errors in the panels)

Actually, there is an encoding bug. I reported at: https://github.com/freeplane/freeplane/issues/1764

**Comment by Olifair** [comment]
this sounds very neat and usefull. And it perhaps also does (or will do in the future) something that I though would be usefull : 

Freeplane is always running, and I usually have many maps open. And when at the end of the day I close FreePlane, it will ask for saving the maps that have changed. Sometimes it will ask for a save for a map for which I do not remember to have changed anything.
But Freeplane will also ask for a save when you only expanded a node, or even clicked on a node. 
Now, I always fear that something might have only accidentally changed in a map. 
So I was looking for a way to see the most recent (10 :-) changes done to the map since the last save. This could assure me that nothing got lost. 
Will this be possible with this add-on ? Automatically ?

**Reply by euu2021** [reply]
> But Freeplane will also ask for a save when you only expanded a node, or even clicked on a node.

Play this setting and tell me if this solves your problem: Preferences…->Environment->Save->Save folding

**Comment by euu2021** [comment]
I included a couple of new features in the script:

- Drag and Drop operations It's possible to do drag and drop (move nodes) between items in the panel, nodes in the map, and vice versa:

![dragff3 borda2](https://github.com/euu2021/Freeplane_UtilityPanels/assets/77707706/e284b3d5-5662-4826-8a49-d37b323578d7)

- Quick Search panel. Quick search that uses the Jumper search engine, so it has transversal search, and other good stuff. I had to adapt Jumper in order to make it accept headless calls.

Results are highlighted everywhere: in the panels, in the inspectors, and in the map:
![image](https://github.com/freeplane/freeplane/assets/77707706/c3754eab-bbda-4a11-acfd-3ce6df372613)

Results auto update when a new match is added in the map:

![quicksearch](https://github.com/euu2021/Freeplane_UtilityPanels/assets/77707706/8f2ad54c-c116-4d46-802b-f02a7e693a0a)


- Infinite inception navigation in the Inspector tooltip:

![inception infinite](https://github.com/euu2021/Freeplane_UtilityPanels/assets/77707706/49e4e946-6ddc-432d-90ce-2f7276d43ced)

- Freeze Inspectors

The inspectors have a freeze option, so the user can easily do drag and drop operations, or navigate the map while the inspectors stay on screen.

![image](https://github.com/euu2021/Freeplane_UtilityPanels/assets/77707706/fc562dfd-f313-4eaf-8cde-ccd253947324)

- pinned nodes are now saved between Freeplane restarts
